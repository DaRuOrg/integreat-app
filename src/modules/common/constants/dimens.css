@custom-media --small-viewport (max-width: 750px);
