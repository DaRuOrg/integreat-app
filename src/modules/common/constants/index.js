import Route from '../../app/Route'
import SearchPage from 'routes/search/containers/SearchPage'
import PdfFetcherPage from 'routes/pdf-fetcher/containers/PdfFetcherPage'
import MainDisclaimerPage from 'routes/main-disclaimer/components/MainDisclaimerPage'
import CategoriesPage from 'routes/categories/containers/CategoriesPage'
import ExtrasPage from 'routes/extras/containers/ExtrasPage'
import DisclaimerPage from 'routes/disclaimer/containers/DisclaimerPage'
import LandingPage from 'routes/landing/containers/LandingPage'
import EventsPage from 'routes/events/containers/EventsPage'

export const LANDING_PAGE_ROUTE = new Route({
  id: LandingPage,
  path: '/(:language(/))'
})
export const SEARCH_PAGE_ROUTE = new Route({
  id: SearchPage,
  path: '/:location/:language/search'
})
export const DISCLAIMER_PAGE_ROUTE = new Route({
  id: DisclaimerPage,
  path: '/:location/:language/disclaimer'
})

export const EVENTS_PAGE_ROUTE = new Route({
  id: EventsPage,
  path: '/:location/:language/events(/:id)'
})

export const PDF_FETCHER_PAGE_ROUTE = new Route({
  id: PdfFetcherPage,
  path: '/:location/:language/fetch-pdf'
})

export const EXTRAS_PAGE_ROUTE = new Route({
  id: ExtrasPage,
  path: '/:location/:language/extras(/:extra(/:type))'
})

export const CATEGORIES_PAGE_ROUTE = new Route({
  id: CategoriesPage,
  path: '/:location/:language(/*)'
})

export const MAIN_DISCLAIMER_PAGE_ROUTE = new Route({
  id: MainDisclaimerPage,
  path: '/disclaimer'
})
