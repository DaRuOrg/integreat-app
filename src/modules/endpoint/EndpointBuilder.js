import { isEqual } from 'lodash/lang'

import Endpoint from './Endpoint'

/**
 * Helper class to build a {@link Endpoint}
 */
class EndpointBuilder {
  _name
  _routerToUrlMapper
  _mapper
  _responseOverride
  _refetchLogic

  /**
   * Creates a new endpoint builder
   * @param {string} name The name of the endpoint to build
   */
  constructor (name) {
    this._name = name

    this._refetchLogic = (urlParams, nextUrlParams) => !isEqual(urlParams, nextUrlParams)
  }

  /**
   * Adds a router to url mapper to the builder
   * @param routerToUrlMapper The url
   * @return {EndpointBuilder} The builder itself
   */
  withRouterToUrlMapper (routerToUrlMapper) {
    this._routerToUrlMapper = routerToUrlMapper
    return this
  }

  /**
   * Adds a url to the builder
   * @param mapper The mapper
   * @return {EndpointBuilder} The builder itself
   */
  withMapper (mapper) {
    this._mapper = mapper
    return this
  }

  /**
   * Adds refetch logic to this builder
   * @param refetchLogic The refetch logic
   * @return {EndpointBuilder} The builder itself
   */
  withRefetchLogic (refetchLogic) {
    this._refetchLogic = refetchLogic
    return this
  }

  /**
   * Overrides value from the API response. Useful for testing.
   * @param responseOverride {*} The response
   * @return {EndpointBuilder} The builder itself
   */
  withResponseOverride (responseOverride) {
    this._responseOverride = responseOverride
    return this
  }

  /**
   * Checks the data and builds the endpoint
   * @return {Endpoint} The final endpoint
   */
  build () {
    if (!this._name) {
      throw Error('You have to set a name to build an endpoint!')
    }

    if (!this._routerToUrlMapper) {
      throw Error('You have to set a url to build an endpoint!')
    }

    if (!this._mapper) {
      throw Error('You have to set a mapper to build an endpoint!')
    }

    if (!this._refetchLogic) {
      throw Error('You have to set a refetch logic to build an endpoint!')
    }

    return new Endpoint(this._name, this._routerToUrlMapper, this._mapper, this._refetchLogic, this._responseOverride)
  }
}

export default EndpointBuilder
