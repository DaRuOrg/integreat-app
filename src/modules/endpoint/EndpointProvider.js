import React from 'react'
import PropTypes from 'prop-types'
import Endpoint from './Endpoint'
import Error from '../common/containers/Error'

class EndpointProvider extends React.Component {
  static propTypes = {
    endpoints: PropTypes.arrayOf(PropTypes.instanceOf(Endpoint)).isRequired,
    children: PropTypes.element.isRequired
  }

  static childContextTypes = {
    getEndpoint: PropTypes.func.isRequired
  }

  getChildContext () {
    return {
      getEndpoint: (endpointName) => {
        const endpoint = this.endpoints[endpointName]
        if (!endpoint) {
          throw new Error('The endpoint ' + endpointName + ' was not found. Maybe you didn\'t wrap the ' +
            'withFetcher(...) call in a <EndpointProvider /> or the endpoint is not registered')
        }
        return endpoint
      }
    }
  }

  constructor (props, context) {
    super(props, context)
    this.endpoints = props.endpoints.reduce((accumulator, endpoint) => {
      accumulator[endpoint.stateName] = endpoint
      return accumulator
    }, {})
  }

  render () {
    return React.Children.only(this.props.children)
  }
}

export default EndpointProvider
