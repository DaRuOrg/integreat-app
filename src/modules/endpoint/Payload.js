/**
 * The payload gets stored in the redux store and holds the information about a etch
 */
class Payload {
  constructor (isFetching, data = null, error = null, requestUrl = null, fetchDate = null) {
    this._isFetching = isFetching
    this._fetchDate = fetchDate
    this._data = data
    this._error = error
    this._requestUrl = requestUrl
  }

  /**
   * @return {number} The date the fetch was initiated as serializable number
   */
  get fetchDate () {
    return this._fetchDate
  }

  /**
   * @return {boolean} If a fetch is going on
   */
  get isFetching () {
    return this._isFetching
  }

  /**
   * @return {*} The data which has been fetched or null
   */
  get data () {
    return this._data
  }

  /**
   * @return {boolean} If the {@link data} is ready to be used
   */
  ready () {
    return !!this._data
  }

  /**
   * @return {string} The error message if the fetch failed or null
   */
  get error () {
    return this._error
  }

  /**
   * @return {string} The url which was used to initiate the fetch
   */
  get requestUrl () {
    return this._requestUrl
  }
}

export default Payload
