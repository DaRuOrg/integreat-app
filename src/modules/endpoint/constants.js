export const apiUrl = 'https://cms.integreat-app.de'
