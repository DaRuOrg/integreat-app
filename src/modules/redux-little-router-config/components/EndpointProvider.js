import React from 'react'
import PropTypes from 'prop-types'
import RouteConfig from '../RouteConfig'

class EndpointProvider extends React.Component {
  static propTypes = {
    routeConfig: PropTypes.instanceOf(RouteConfig).isRequired,
    children: PropTypes.element.isRequired
  }

  static childContextTypes = {
    matchRoute: PropTypes.func.isRequired
  }

  getChildContext () {
    return {
      matchRoute: (id) => this.props.routeConfig.matchRoute(id)
    }
  }

  render () {
    return React.Children.only(this.props.children)
  }
}

export default EndpointProvider
