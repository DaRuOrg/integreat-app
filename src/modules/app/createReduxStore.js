// @flow

import type { Store } from 'redux'
import { applyMiddleware, combineReducers, compose, createStore } from 'redux'
import { createLogger } from 'redux-logger'

import { languageReducer, uiDirectionReducer } from '../i18n/reducers'
import toggleDarkModeReducer from '../theme/reducers'
import Payload from '../endpoint/Payload'

export type ActionType<T> = { type: string, payload: Payload<T> }

// todo: Change type to correct State type,
// https://blog.callstack.io/type-checking-react-and-redux-thunk-with-flow-part-2-206ce5f6e705
const createReduxStore = (initialState: {} = {}): Store<any, any> => {
  /**
   * The middlewares of this app, add additional middlewares here
   */
  const middlewares = []

  if (__DEV__) {
    middlewares.push(createLogger()) // Logs all state changes in console
  }

  const rootReducer = combineReducers({
    uiDirection: uiDirectionReducer,
    language: languageReducer,
    darkMode: toggleDarkModeReducer
  })

  const enhancers = compose(applyMiddleware(...middlewares))

  return createStore(rootReducer, initialState, enhancers)
}

export default createReduxStore
