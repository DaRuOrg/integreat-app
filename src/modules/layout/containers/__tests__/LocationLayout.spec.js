import React from 'react'
import { Provider } from 'react-redux'
import EndpointBuilder from 'modules/endpoint/EndpointBuilder'
import { mount, shallow } from 'enzyme'
import ConnectedLocationLayout, { LocationLayout } from '../LocationLayout'
import LocationModel from 'modules/endpoint/models/LocationModel'
import EndpointProvider from '../../../endpoint/EndpointProvider'
import { HALF_HEADER_HEIGHT_SMALL, HEADER_HEIGHT_LARGE } from '../../constants'
import createReduxStore from '../../../app/createReduxStore'
import createHistory from '../../../app/createHistory'
import thunk from 'redux-thunk'
import configureMockStore from 'redux-mock-store'
import Payload from '../../../endpoint/Payload'

describe('LocationLayout', () => {
  const matchRoute = (id) => {}

  const language = 'de'

  const locations = [new LocationModel({name: 'Mambo No. 5', code: 'location1'})]

  const MockNode = () => <div />

  test('should show LocationHeader and LocationFooter if LocationModel is available', () => {
    const component = shallow(
      <LocationLayout location='location1' language={language}
                      matchRoute={matchRoute}
                      locations={locations}
                      scrollHeight={0}
                      currentPath='/:location/:language'>
        <MockNode />
      </LocationLayout>)
    expect(component).toMatchSnapshot()
  })

  test('should show GeneralHeader and GeneralFooter if LocationModel is not available', () => {
    const component = shallow(
      <LocationLayout location='unavailableLocation' language={language}
                      matchRoute={matchRoute}
                      locations={locations}
                      scrollHeight={0}
                      currentPath='/:location/:language'>
        <MockNode />
      </LocationLayout>)
    expect(component).toMatchSnapshot()
  })

  describe('connect()', () => {
    const locationsEndpoint = new EndpointBuilder('locations')
      .withStateToUrlMapper(() => 'https://weird-endpoint/api.json')
      .withMapper(json => json)
      .withResponseOverride(locations)
      .build()

    const location = 'augsburg'
    const path = '/:location/:language'

    test('should map state to props', () => {
      const store = createReduxStore(createHistory, {
        router: {params: {location: location, language: language}, route: path},
        viewport: {is: {small: false}}
      })

      const locationLayout = mount(
        <Provider store={store}>
          <EndpointProvider endpoints={[locationsEndpoint]}>
            <ConnectedLocationLayout />
          </EndpointProvider>
        </Provider>
      ).find(LocationLayout)

      expect(locationLayout.props()).toEqual({
        currentPath: path,
        location: location,
        language: language,
        locations: locations,
        scrollHeight: HEADER_HEIGHT_LARGE,
        dispatch: expect.any(Function)
      })
    })

    const mockStore = configureMockStore([thunk])

    const createComponentInViewport = (small) => {
      const smallStore = mockStore({
        locations: new Payload(false),
        router: {params: {location: 'augsburg', language: 'en', id: '1234'}, route: '/:location/:language'},
        viewport: {is: {small}}
      })
      return mount(
        <Provider store={smallStore}>
          <EndpointProvider endpoints={[locationsEndpoint]}>
            <ConnectedLocationLayout><MockNode /></ConnectedLocationLayout>
          </EndpointProvider>
        </Provider>
      )
    }

    test('should have correct scroll height', () => {
      const smallComponent = createComponentInViewport(true).find(ConnectedLocationLayout).childAt(0)
      expect(smallComponent.prop('scrollHeight')).toBe(HALF_HEADER_HEIGHT_SMALL)

      const largeComponent = createComponentInViewport(false).find(ConnectedLocationLayout).childAt(0)
      expect(largeComponent.prop('scrollHeight')).toBe(HEADER_HEIGHT_LARGE)
    })
  })
})
