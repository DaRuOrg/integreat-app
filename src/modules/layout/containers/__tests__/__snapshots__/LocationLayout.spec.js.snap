// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationLayout connect should map state to props 1`] = `
Object {
  "children": <MockNode />,
  "dispatch": [Function],
  "location": "augsburg",
  "navigation": Navigation {
    "_language": "en",
    "_location": "augsburg",
  },
  "route": "/:location/:language",
}
`;

exports[`LocationLayout should show GeneralHeader and GeneralFooter if LocationModel is not available 1`] = `
<Layout
  footer={<Translate(GeneralFooter) />}
  header={<GeneralHeader />}
>
  <MockNode />
</Layout>
`;

exports[`LocationLayout should show LocationHeader and LocationFooter if LocationModel is available 1`] = `
<Layout
  footer={
    <Translate(LocationFooter)
      navigation={
        Navigation {
          "_language": "language1",
          "_location": "location1",
        }
      }
    />
  }
  header={
    <Translate(LocationHeader)
      location={
        LocationModel {
          "_code": "location1",
          "_eventsEnabled": undefined,
          "_extrasEnabled": undefined,
          "_live": undefined,
          "_name": "Mambo No. 5",
          "_sortCategory": "M",
          "_sortKey": "mambo no. 5",
        }
      }
      navigation={
        Navigation {
          "_language": "language1",
          "_location": "location1",
        }
      }
      route="/:location/:language"
    />
  }
>
  <MockNode />
</Layout>
`;
